### ISSUE TITLE
Linking Programs Compiled from Source with Programs Compiled from IL

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#676]
Creating this issue from a discussion in a SPIRV-LLVM-Translator PR review:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1149#discussion_r707579538

Is it possible to link one program object created from source with another program object created from IL?

Discussion "for":

> The program object created from source or IL must be compiled (via clCompileProgram) before it can be linked, which transitions the state of the program object from BINARY_TYPE_NONE to BINARY_TYPE_COMPILED_OBJECT. After the program object has been compiled I can't think of any reason to differentiate between program objects that originated from source and program objects that originate from IL, so the linking should "just work". Is there a detail I'm missing here?

Discussion "against":

> C++ compilers have similar problems here, where a "compiled object" from one version of a compiler doesn't necessarily have the same ABI as a "compiled object" from another version of a compiler, and therefore the two objects can't necessarily be linked. This is similar, if we treat an IL program as an object from one version of a compiler, which might not be the same as the platform's source compiler.

**Comment by StuartDBrady** [comment]
> Discussion "for":
> 
> > The program object created from source or IL must be compiled (via clCompileProgram) before it can be linked, which transitions the state of the program object from BINARY_TYPE_NONE to BINARY_TYPE_COMPILED_OBJECT. After the program object has been compiled I can't think of any reason to differentiate between program objects that originated from source and program objects that originate from IL, so the linking should "just work". Is there a detail I'm missing here?

Indeed, I can't see what error code the `clLinkProgram()` is _supposed_ to return when attempting to link source with IL – therefore, I assume it is perfectly valid to do this, and any implementation that does not allow it is incomplete.

> Discussion "against":
> 
> > C++ compilers have similar problems here, where a "compiled object" from one version of a compiler doesn't necessarily have the same ABI as a "compiled object" from another version of a compiler, and therefore the two objects can't necessarily be linked. This is similar, if we treat an IL program as an object from one version of a compiler, which might not be the same as the platform's source compiler.

This applies even when linking multiple SPIR-V modules _without_ linking with any programs compiled from source. Each language targeting SPIR-V needs to define a SPIR-V ABI. I believe for C++ for OpenCL, we essentially follow the Itanium C++ ABI.

If additional languages are supported using `clCreateProgramWithSource()` through a CL extension, devices that support `clCreateProgramWithIL()` will require that the online compilation also be compatible with the language's SPIR-V ABI.

**Comment by StuartDBrady** [comment]
Note, as commented on issue #675, I cannot see where we have even defined a SPIR-V ABI for OpenCL C.
